src/server/worker/datum/iterator.go

```diff
@@ -444,18 +444,15 @@ func (ds *datumStream) Next() error {
 			return io.EOF
 		}
 		ds.meta = meta
+		ds.id = common.DatumID(meta.Inputs)
 		return nil
 	case err := <-ds.errChan:
 		return err
 	}
 }
 
-func (ds *datumStream) Key() string {
-	return common.DatumID(ds.meta.Inputs)
-}
-
-func (ds *datumStream) Priority() int {
-	return ds.priority
+func (ds *datumStream) Compare(s stream.Stream) int {
+	return strings.Compare(ds.id, s.(*datumStream).id)
```

Comment:
This looks like it could panic if it is given something that's not a datumStream.

It's usually better to have a comparison function (like the sort package).  Because it will actually be in a position to know that it will never be comparing non datumStreams.  There is no way of knowing that from the context of this method.

Reply:
Yeah, that can happen in the callback as well. I think the streaming functionality may make moving to a model like that add more complexity than value since the client would need to manage a mapping from indexes to a set of streams. We could add a type check to the constructor to ensure that all of the stream are of the same type if you are concerned about this? I am up for changes to the priority queue if you have a cleaner model in mind though.

Reply:
My recommendation is to pass a compare function to the priority queue.  so a `func(a, b Stream) int`.  This is basically the same thing as the current compare method.  However, this will be defined as a lambda in practice, and it will be defined right next to the `streams := []Stream{s1, s2, s3}` .  So the lambda in context will know that all the streams can be cast safely to whatever concrete type.  A public method `Compare(other Stream)`  on the other hand, just assumes it will never compare anything else, which is not a great assumption considering that it’s public.

This is pretty much the same reason why the newer sort methods like Slice take a comparison function, but some older data structure packages define interfaces that do a similar cast to their own concrete type.